libsolidity/analysis/ReferencesResolver.cpp

```diff
@@ -367,8 +367,29 @@ void ReferencesResolver::endVisit(VariableDeclaration const& _variable)
 				}
 				else if (varLoc == Location::Default)
 				{
-					if (_variable.isCallableParameter())
-						typeLoc = DataLocation::Memory;
+					if (_variable.isCallableParameter()) {
+                        typeLoc = DataLocation::Memory;
+
+                        // Warn users (error in 0.5.0) for using default data location in libraries.
+                        auto const& contract = dynamic_cast<ContractDefinition const&>(
+                                *dynamic_cast<Declaration const&>(*_variable.scope()).scope()
+                        );
+                        if (contract.isLibrary()) {
```

Comment:
Why should this be restricted to libraries only?

Note that for `external` functions, we have a default data location of `calldata` which is not possible to express in source code, so we cannot warn about it.

Reply:
In the original issue you mentioned "We should do the same at least for the case where the contract is a library.", so I only warned for libraries, but I have changed it to warn for contracts as well.